File: day48/reverseBetween.java
/*
反转从位置 m 到 n 的链表。请使用一趟扫描完成反转。
1 ≤ m ≤ n ≤ 链表长度。

输入: 1->2->3->4->5->NULL, m = 2, n = 4
输出: 1->4->3->2->5->NULL

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/reverse-linked-list-ii
 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode res = new ListNode(0);
        res.next = head;
        ListNode pivot = res;

        for (int i = 1; i < m; i++) {
            pivot = pivot.next;
        }
        
        ListNode cur = pivot.next;
        for (int i = m; i < n; i++) {
            ListNode temp = cur.next;
            cur.next = temp.next;
            temp.next = pivot.next;
            pivot.next = temp;
        }

        return res.next;
    }
}

File: day48/reverseBetween.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        res = ListNode(0)
        res.next = head
        pivot = res
        for _ in range(m - 1):
            pivot = pivot.next

        cur = pivot.next
        for _ in range(n - m):
            temp = cur.next
            cur.next = temp.next
            temp.next = pivot.next
            pivot.next = temp

        return res.next
